internal/cloudsql/instance.go

```diff
@@ -0,0 +1,192 @@
+	case <-i.ready:
+		return i.err
+	}
+}
+
+// Instance manages the information used to connect to the Cloud SQL instance by periodically calling
+// the Cloud SQL Admin API. It automatically refreshes the required information every 55 minutes.
+type Instance struct {
+	connName
+	client *sqladmin.Service
+	key    *rsa.PrivateKey
+
+	resultGuard *sync.RWMutex
+	cur         *refreshResult
+	next        *refreshResult
+
+	// TODO: add a way to close
+}
+
+// NewInstance initializes a new Instance given an instance conneciton name
+func NewInstance(instanceConnName string, client *sqladmin.Service, key *rsa.PrivateKey) (*Instance, error) {
+	cn, err := parseConnName(instanceConnName)
+	if err != nil {
+		return nil, err
+	}
+	i := &Instance{cn, client, key, &sync.RWMutex{}, nil, nil}
+	// Kick off the inital refresh op asynchronously
+	i.resultGuard.Lock()
+	i.cur = i.scheduleRefresh(0)
+	i.next = i.cur
+	i.resultGuard.Unlock()
+	return i, nil
+}
+
+// Connect returns a secure, authorized net.Conn to a Cloud SQL instance.
+func (im *Instance) Connect(ctx context.Context) (net.Conn, error) {
+	im.resultGuard.RLock()
+	res := im.cur
+	im.resultGuard.RUnlock()
+	err := res.Wait(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Add better ipType support, including an opt to specify.
+	addr := net.JoinHostPort(res.md.ipAddrs["PUBLIC"], "3307")
+	conn, err := proxy.Dial(ctx, "tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	if c, ok := conn.(*net.TCPConn); ok {
+		if err := c.SetKeepAlive(true); err != nil {
+			return nil, fmt.Errorf("failed to set keep-alive: %w", err)
+		}
+		if err := c.SetKeepAlivePeriod(30 * time.Second); err != nil {
+			return nil, fmt.Errorf("failed to set keep-alive period: %w", err)
+		}
+	}
+
+	tlsConn := tls.Client(conn, res.tlsCfg)
+	if err := tlsConn.Handshake(); err != nil {
+		tlsConn.Close()
+		return nil, fmt.Errorf("handshake failed: %w", err)
+	}
+	return tlsConn, err
+}
+
+// scheduleRefresh schedules a refresh operation to be triggered after a given duration. The returned resultRefresh
+// can be used to either Cancel or Wait for the operations result.
+func (im *Instance) scheduleRefresh(d time.Duration) *refreshResult {
+	res := &refreshResult{}
+	res.ready = make(chan struct{})
+	res.timer = time.AfterFunc(d, func() {
+		performRefresh(*im, res, d)
+		// Once the refresh has been performed, replace "current" with the most recent result and schedule a new refresh
+		im.resultGuard.Lock()
+		if res.err == nil {
+			im.cur = res
+			im.next = im.scheduleRefresh(55 * time.Minute)
+		} else {
+			// If something went wrong, schedule the next refresh immediately instead
+			// TODO: only replace cur result if it's still valid
+			im.cur = res
+			im.next = im.scheduleRefresh(0)
+		}
+		im.resultGuard.Unlock()
+	})
+	return res
+}
+
+// performRefresh immediately perfoms a full refresh operation using the Cloud SQL Admin API.
+func performRefresh(im Instance, res *refreshResult, d time.Duration) {
```

Comment:
Yes, you're correct. I'm thinking of `refreshResult`, but it's not clear that adding this function as a method on that type would be much better. Still it seems a little awkward to pass in a pointer to refreshResult, mutate it, and not return anything.

Reply:
Yeah I agree with what you're saying. The alternative I considered was to use something like `refreshResult.New(i *Instance, d time.Duration)`, but then I was still passing a pointer. It seemed cleaner to do it this way, but I could probably be persuaded otherwise.